src/github.com/getlantern/enproxy/conn_test.go

```diff
@@ -62,18 +60,80 @@ func TestBadBuffered(t *testing.T) {
 	doTestBad(true, t)
 }
 
+func TestIdle(t *testing.T) {
+	idleTimeout := 100 * time.Millisecond
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	_, err = Dial(httpAddr, &Config{
+		DialProxy: func(addr string) (net.Conn, error) {
+			return net.Dial("tcp", proxyAddr)
+		},
+		NewRequest:  newRequest,
+		IdleTimeout: idleTimeout,
+	})
+	if assert.NoError(t, err, "Dialing should have succeeded") {
+		time.Sleep(idleTimeout * 2)
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+	}
+}
+
+// This test stimulates a connection leak as seen in
+// https://github.com/getlantern/lantern/issues/2174.
+func TestHTTPRedirect(t *testing.T) {
+	startProxy(t)
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			Dial: func(network, addr string) (net.Conn, error) {
+				return Dial(addr, &Config{
+					DialProxy: func(addr string) (net.Conn, error) {
+						return net.Dial("tcp", proxyAddr)
+					},
+					NewRequest: newRequest,
+				})
+			},
+			DisableKeepAlives: true,
+		},
+	}
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	resp, err := client.Head("http://www.facebook.com")
+	if assert.NoError(t, err, "Head request to facebook should have succeeded") {
+		resp.Body.Close()
+	}
+
+	assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+}
+
 func doTestPlainText(buffered bool, t *testing.T) {
 	startServers(t)
 
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
 	conn, err := prepareConn(httpAddr, buffered, false, t)
 	if err != nil {
 		t.Fatalf("Unable to prepareConn: %s", err)
 	}
-	defer conn.Close()
+	defer func() {
+		err := conn.Close()
+		assert.Nil(t, err, "Closing conn should succeed")
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
```

Comment:
I've tracked down the mysterious opened-then-closed connection that caused our `CLOSE_WAIT` to the `WaitForServer()` function within the `waitforserver` package. The reason is this function attempts to dial to a network address and, if successful, it immediately quits. So the `CLOSE_WAIT` origin is now clear (and it's also completely expected).

The `WaitForServer()` function is mostly used in tests with the exception of `./github.com/getlantern/yamlconf/yamlconf_http.go`. I'd suggest to change the function a bit to allow the `CLOSE_WAIT` status to disappear. I suggest changing the `WaitForServer()` function directly since this change would be practically inoffensive and would avoid changing other parts of our code to add a timer.

```
func WaitForServer(protocol string, addr string, limit time.Duration) error {
  defer time.Sleep(50 * time.Millisecond)

  cutoff := time.Now().Add(limit)
  for {
    if time.Now().After(cutoff) {
      return fmt.Errorf("Server never came up at %s address %s", protocol, addr)
    }   
    c, err := net.DialTimeout(protocol, addr, limit)
    if err == nil {
      return c.Close()
    }   
    time.Sleep(50 * time.Millisecond)
  }
}
```

Another solution would be adding a `time.Sleep(time.Millisecond*100)` after the `startServers()` or after any occurrence of the `WaitForServer()` function.